modules/lightmapper_rd/lm_compute.glsl

```diff
@@ -461,7 +461,7 @@ void trace_direct_light(vec3 p_position, vec3 p_normal, uint p_light_index, bool
 #if defined(MODE_BOUNCE_LIGHT) || defined(MODE_LIGHT_PROBES)
 
 vec3 trace_environment_color(vec3 ray_dir) {
-	vec3 sky_dir = normalize(mat3(bake_params.env_transform) * ray_dir);
+	vec3 sky_dir = normalize(ray_dir * mat3(bake_params.env_transform));
```

Comment:
This looks suspicious. This makes it seem like `env_transform` is transposed for whatever reason. I wonder if `from_euler()` is giving us a transposed matrix or something

Reply:
Ah, look what I found in the rendering code, we always use the inverse:
https://github.com/godotengine/godot/blob/607b230ffe120b2757c56bd3d52a7a0d4e502cfe/servers/rendering/renderer_rd/environment/sky.cpp#L1543-L1544

In this case, it works because the basis is an orthogonal matrix, which means that the inverse is the same as the transpose. Switching the order of operations works here because you can reverse the order of operations instead of transposing a matrix.

The correct fix here would be to revert this change and invert `environment_transform` before passing it to the shader.

Reply:
This PR is originally my code, so this is my bad. At the time, I managed to convince myself that reversing the order here made sense. Thinking about it now, I see that it doesn't.

> The correct fix here would be to revert this change and invert environment_transform before passing it to the shader.

Even better, we can revert this change and remove the invert. `sky_transform` isn't used for anything else so inverting it twice is redundant. Why is it inverted in the first place?

Reply:
> Even better, we can revert this change and remove the invert. `sky_transform` isn't used for anything else so inverting it twice is redundant. Why is it inverted in the first place?

OK, now I'm just being stupid. Please ignore.

Reply:
The PR can be reduced to a single line:

```diff
--- a/scene/3d/lightmap_gi.cpp
+++ b/scene/3d/lightmap_gi.cpp
@@ -1072,6 +1072,7 @@ LightmapGI::BakeError LightmapGI::bake(Node *p_from_node, String p_image_data_pa
 
 					if (env.is_valid()) {
 						environment_image = RS::get_singleton()->environment_bake_panorama(env->get_rid(), true, Size2i(128, 64));
+						environment_transform = Basis::from_euler(env->get_sky_rotation()).inverse();
 					}
 				}
 			} break;
```

Is the reason for having to invert here that the shader is taking the perspective of not actually rotating the sky but instead keeping the sky fixed and rotating everything else? My head hurts.

Reply:
> Is the reason for having to invert here that the shader is taking the perspective of not actually rotating the sky but instead keeping the sky fixed and rotating everything else? My head hurts.

Honestly, I'm not sure. I would have to really dig in to the math to figure it out.

My gut feeling is that `from_euler` returns a transposed matrix (either returning column order when we expect row order or vice versa). And we are inverting the basis as a way of getting the transposed matrix.

Reply:
I tried hardcoding a basis using `Basis().rotated()` and it still had to be inverted to be correct, so I don't think `from_euler` is to blame.

I should let this go, but I'm starting to come full circle and wonder if my original change to the shader wasn't fine after all. We have a world space vector (`ray_dir`) that needs to be made local to the sky's transform, which is what `ray_dir * transform` would do.

Anyway, the one-liner in the diff above with the invert works and is consistent with how the sky rotation is used in another place, so let's go with that.

Reply:
What should I change in my PR in the end?

Reply:
You can do the change suggested by permelin in this comment https://github.com/godotengine/godot/pull/95000#discussion_r1700071873

Reply:
I've tested that change and it works, thanks :slightly_smiling_face: